irc/metadata.go

```diff
@@ -0,0 +1,257 @@
+package irc
+
+import (
+	"errors"
+	"regexp"
+	"slices"
+	"strings"
+
+	"maps"
+
+	"github.com/ergochat/ergo/irc/caps"
+	"github.com/ergochat/ergo/irc/modes"
+	"github.com/ergochat/ergo/irc/utils"
+)
+
+var (
+	errMetadataTooManySubs = errors.New("too many subscriptions")
+)
+
+type MetadataStore = map[string]string
+
+type MetadataThing = struct {
+	Target     string
+	Key        string
+	Visibility string // currently unused
+	Value      string
+}
+
+func metadataSet(server *Server, session *Session, target string, key string, value string) {
+	targetChannel := server.channels.Get(target)
+	targetClient := server.clients.Get(target)
+
+	if targetChannel != nil {
+		targetChannel.stateMutex.Lock()
+
+		targetChannel.metadata[key] = value
+		targetChannel.stateMutex.Unlock()
+		targetChannel.MarkDirty(IncludeAllAttrs)
+		return
+	}
+
+	if targetClient != nil {
+		targetClient.stateMutex.Lock()
+		defer targetClient.stateMutex.Unlock()
+
+		if targetClient.metadata == nil {
+			targetClient.metadata = make(MetadataStore)
+		}
+
+		targetClient.metadata[key] = value
+
+		// coming soon: https://www.youtube.com/watch?v=K14JkFfWUzc
+
+		return
+	}
+
+	notifySubscribers(server, session, target, key, value)
```

Comment:
You're still holding `stateMutex` here which is problematic...I would actually move the underlying get/set/delete operations to methods in `irc/getters.go` which take the lock, do the operation, release the lock, and return a success status if necessary (see `(*Client).deletePushSubscription` for example). And then you can send messages while not holding the lock.